// gatsby-source-wordpress/dist/models/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.settings = exports.default = void 0;
var _remoteSchema = _interopRequireDefault(require("./remoteSchema"));
var _gatsbyApi = _interopRequireDefault(require("./gatsby-api"));
var _logger = _interopRequireDefault(require("./logger"));
var _imageNodes = _interopRequireDefault(require("./image-nodes"));
var _wpHooks = _interopRequireDefault(require("./wp-hooks"));
var _preview = _interopRequireDefault(require("./preview"));
var _develop = _interopRequireDefault(require("./develop"));
var _postBuildWarningLogs = _interopRequireDefault(require("./post-build-warning-logs"));
var _core = require("@rematch/core");
const settings = (0, _core.createModel)()({
  state: 0,
  reducers: {
    increment: (state, payload) => state + payload
  },
  effects: () => {
    return {};
  }
});
exports.settings = settings;
const models = {
  remoteSchema: _remoteSchema.default,
  gatsbyApi: _gatsbyApi.default,
  logger: _logger.default,
  imageNodes: _imageNodes.default,
  wpHooks: _wpHooks.default,
  previewStore: _preview.default,
  develop: _develop.default,
  postBuildWarningCounts: _postBuildWarningLogs.default
};
var _default = models;
exports.default = _default;
